### PR TITLE
[Sol - Yul] Add support for built-in logN().

### PR DESCRIPTION
**Pull request by aarlt**
Part of #8343.

### CODE REVIEW

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 1]
```diff
@@ -812,6 +812,35 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
...
+	case FunctionType::Kind::Log2:
+	case FunctionType::Kind::Log3:
+	case FunctionType::Kind::Log4:
+	{
+		unsigned logNumber = int(functionType->kind()) - int(FunctionType::Kind::Log0);
+		solAssert(arguments.size() == logNumber + 1, "");
+		ABIFunctions abi(m_context.evmVersion(), m_context.revertStrings(), m_context.functionCollector());
+		vector<IRVariable> indexedArgs;
+		for (unsigned arg = 0; arg < logNumber; ++arg)
+			indexedArgs.emplace_back(convert(*arguments[arg + 1], *(parameterTypes[arg + 1])));
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: `expressionAsType`?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp [hunk 2]
```diff
@@ -812,6 +812,35 @@ void IRGeneratorForStatements::endVisit(FunctionCall const& _functionCall)
...
+			<log>(<pos>, sub(<end>, <pos>) <indexedArgs>)
+		})");
+		templ("pos", m_context.newYulVariable());
+		templ("end", m_context.newYulVariable());
+		templ("freeMemory", freeMemory());
+		templ("encode", abi.tupleEncoder({arguments.front()->annotation().type},{parameterTypes.front()}));
+		templ("nonIndexedArgs", IRVariable(*arguments.front()).commaSeparatedList());
+		templ("log", "log" + to_string(logNumber));
+		templ("indexedArgs", joinHumanReadablePrefixed(indexedArgs | boost::adaptors::transformed([&](auto const& _arg) {
+			return _arg.commaSeparatedList();
```
Comment (chriseth) [inline libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: Why not already do this in the loop above?

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 1 above)
Comment (aarlt) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: may be better

File: libsolidity/codegen/ir/IRGeneratorForStatements.cpp (hunk 2 above)
Comment (aarlt) [reply libsolidity/codegen/ir/IRGeneratorForStatements.cpp]: thats better :)

**Review by chriseth — APPROVED** [review]

**Pull request merged**